tests/test_constant_visitor.py

```diff
@@ -22,66 +22,74 @@
+        self.assertEqual({}, tree.accept(visitor))
         tree = BooleanConstant('c0')
-        constants = tree.accept(ConstantVisitor())
-        self.assertEqual(constants, {'c0': 'Bool'})
+        self.assertEqual({'c0': 'Bool'}, tree.accept(visitor))
         tree = BooleanEquality(BooleanConstant('c0'), BooleanConstant('c1'))
-        constants = tree.accept(ConstantVisitor())
-        self.assertEqual(constants, {'c0': 'Bool', 'c1': 'Bool'})
+        self.assertEqual({'c0': 'Bool', 'c1': 'Bool'}, tree.accept(visitor))
 
     def test_integer_visitor_easy(self):
+        visitor = ConstantVisitor()
         tree = IntegerVariable('x')
-        constants = tree.accept(ConstantVisitor())
-        self.assertEqual(constants, {})
+        self.assertEqual({}, tree.accept(visitor))
         tree = IntegerConstant('c0')
-        constants = tree.accept(ConstantVisitor())
-        self.assertEqual(constants, {'c0': 'Int'})
+        self.assertEqual({'c0': 'Int'}, tree.accept(visitor))
         tree = IntegerEquality(IntegerConstant('c0'), IntegerConstant('c1'))
-        constants = tree.accept(ConstantVisitor())
-        self.assertEqual(constants, {'c0': 'Int', 'c1': 'Int'})
+        self.assertEqual({'c0': 'Int', 'c1': 'Int'}, tree.accept(visitor))
 
     def test_real_visitor_easy(self):
-        tree = RealVariable('c0')
-        constants = tree.accept(ConstantVisitor())
-        self.assertEqual(constants, {})
+        visitor = ConstantVisitor()
+        tree = RealVariable('x')
+        self.assertEqual({}, tree.accept(visitor))
         tree = RealConstant('c0')
-        constants = tree.accept(ConstantVisitor())
-        self.assertEqual(constants, {'c0': 'Real'})
+        self.assertEqual({'c0': 'Real'}, tree.accept(visitor))
         tree = RealEquality(RealConstant('c0'), RealConstant('c1'))
-        constants = tree.accept(ConstantVisitor())
-        self.assertEqual(constants, {'c0': 'Real', 'c1': 'Real'})
+        self.assertEqual({'c0': 'Real', 'c1': 'Real'}, tree.accept(visitor))
+
+    def test_string_visitor_easy(self):
+        visitor = ConstantVisitor()
+        tree = StringVariable('x')
+        self.assertEqual({}, tree.accept(visitor))
+        tree = StringLiteral('c0')
+        self.assertEqual({'c0': 'String'}, tree.accept(visitor))
+        tree = StringEquality(StringLiteral('c0'), StringLiteral('c1'))
+        self.assertEqual({'c0': 'String', 'c1': 'String'}, tree.accept(visitor))
 
     def test_boolean_visitor_inequality(self):
-        tree = BooleanEquality(BooleanConstant('c0'), BooleanConstant('c1'))
-        constants = tree.accept(ConstantVisitor())
-        self.assertNotEqual(constants, {'c0': 'Boolean', 'c1': 'Bool', 'c2': 'Bool'})
+        visitor = ConstantVisitor()
+        tree_1 = BooleanEquality(BooleanConstant('c0'), BooleanConstant('c1'))
+        tree_2 = BooleanEquality(BooleanConstant('c1'), BooleanConstant('c2'))
+        self.assertNotEqual(tree_1.accept(visitor), tree_2.accept(visitor))
```

Comment:
Oh so now I get what you were saying, sorry but I did forget about the precise implementation. I am not sure whether of the two is more effective, so both are fine for me